src/Sarif/ExtensionMethods.cs

```diff
@@ -40,6 +40,11 @@ public static bool HasAtLeastOneNonNullValue<T>(this IEnumerable<T> collection)
             return collection != null && collection.Any((m) => m != null);
         }
 
+        public static bool HasAtLeastOneNonNullEmptyValue(this MultiformatMessageString multiformatMessageString)
```

Comment:
>HasAtLeastOneNonNullEmptyValue [](start = 27, length = 30)

This isn't quite right, I think. Your serialization logic would fail to emit a property value with a multiformat message string that had a markdown property but no text value. This isn't valid SARIF but the point here is not to ensure valid SARIF by dropping explicitly provided data, make sense? Instead, it is here to drop verifiably empty objects. MutiformatMessageString is a lightweight object, so you can simply add checks to see whether the Markdown property is available, and/or the properties bag is non-null with a count of items that exceeds zero. I suggest adding tests first that would have caught these removals (i.e., try to serialize a multiformat message string with only a markdown value and author a test that verifies it survives. same for the property bag only scenario).

Reply:
Changing locally as you suggested:

the section below is the rules inside a sarif to test if we can serialize/deserialize successfully
```csharp
Rules = new List<ReportingDescriptor>
                                {
                                    new ReportingDescriptor
                                    {
                                        FullDescription = new MultiformatMessageString
                                        {
                                            Markdown = "markdown text"
                                        }
                                    }
                                }
```

it generates after the first serialization the following json:
```json
{
                            "fullDescription": {
                                "markdown": "markdown text",
                            }
                        }
```

when i try to deserialize to sariflog, it throws an exception:
`  Newtonsoft.Json.JsonSerializationException : Required property 'text' not found in JSON. Path 'runs[0].tool.driver.rules[0].fullDescription', line 1, position 209.`

if we just add the code that i implemented, the fullDescription wouldn't be serialized, so it would be wrong as well. But, if we serialize, we have to add somehow the text to be serializable again....

What do you think?


---
In reply to: [440881150](https://github.com/microsoft/sarif-sdk/pull/1885#discussion_r440881150) [](ancestors = 440881150)

Reply:
I see. So we have dueling values here. On the one hand, we don't want people to be able to serialize SARIF which subsequently doesn't deserialize. On the other hand, we don't want to unexpectedly remove any data in order to ensure what's persisted to disk is valid. @lgolding do you have any thoughts on this?

---
In reply to: [444911014](https://github.com/microsoft/sarif-sdk/pull/1885#discussion_r444911014) [](ancestors = 444911014,440881150)